### ISSUE TITLE
Analysis server returns duplicate search results (using flutter repo root as analysis root)

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#31589]
I noticed this in search results, though suspect it may affect other places too.

If you send multiple analysis roots that include the same files (in VS Code, it's reasonable to have child projects open as additional roots for launching/debugging purposes), for example:

- /dev/flutter
- /dev/flutter/examples/stocks
- /dev/flutter/examples/flutter_gallery

If you send a `findTopLevelDeclerations` for something that exists in both, you'll get duplicates in the results (it's also possible that resources were spent searching the files twice, I'm not sure).

As a workaround, I'm going to filter out any roots that are descendants or other roots but @bwilkerson thinks this would be best handled in the analysis server ultimately.

**Comment by DanTup** [comment]
@bwilkerson Ok, I don't think this is related to overlapping roots at all! I implemented a workaround and found the issue still persisted. I did some more testing and found that:

- If I open `/dev/flutter/examples/stocks` as a root and search for `StockArrowPainter` I get a single result
- If I open `/dev/flutter` as a root and search for `StockArrowPainter` I get the same result duplicated

[Log file](https://github.com/dart-lang/sdk/files/1544701/analyzer.txt)

I'm not sure what's causing this (or whether it's something weird about the flutter repo, it might be common in others) but I'm going to revert my workaround for now (since it doesn't work).
